readme.txt
+-----------------------------------------------------+
+++++++++++++++++++++ WebAuthor +++++++++++++++++++++++
+--------------Maintained By CI Software--------------+
+-----------------------------------------------------+

About:
The intuitive, open source webpage creation environment.

Contact:
    Charles Ivan Mozar ( CI SOftware )
        --> charlesivanmozar.cisoftware@gmail.com


How to Compile:
    > Download Qt ( open source ) from http://qt.io
    > then open the WebAuthor project file

    (NOTE: for OS X / MacOS, you may need XCode
           you can download it from the Mac App Store for free)

+-------------------------------------------------------+
+----------------Source file information----------------+
+-------------------------------------------------------+

webauthor.h / webauthor.cpp: (03/07/17)
 --> This is the main window class, everything is initialized
      here

wbhtmlsyntax.h / wbhtmlsyntax.cpp (03/08/17)
 --> This class handles the HTML syntax highlighting system

wbcsssyntax.h / wbcsssyntax.cpp (03/08/17)
 --> This class handles the Cascading Style Sheet highlighting
        system

wbjssyntax.h / wbjssyntax.cpp (03/08/17)
 --> This class handles the Javascript syntax highlighting system

wbcolorlistmodel.h / wbcolorlistmodel.cpp (03/09/17)
 --> This creates the abstract model class for the color picker tool

wblibrarian.h / wblibrarian.cpp (03/11/17)
 --> This implements the "Librarian" widget ( a dictionary for html and others )

wblibrarianmodel.h / wblibrarianmodel.cpp (03/11/17)
 --> Implements the Librarian ListView model ( also the loading of .wblibrarian files )

wbcodecompletionmodel.h / wbcodecompletionmodel.cpp (03/10/17)
 --> This creates the abstract model class for code completion

wbcodeview.h / wbcodeview.cpp (03/10/17)
 --> This defines the custom QTextEdit class we use for the editors

----- On the resource.qrc file -----

gettingstarted.htm (03/07/17)
 --> Contains the "Getting Started" page that is shown when the user
     presses the F1 Key

icon.png (03/07/17)
 --> WebAuthor's icon

html_elements.wblist (03/10/17)
 --> contains the HTML autocomplete data

css_elements.wblist (03/10/17)
 --> contains the Cascading Style Sheets autocomplete data

js_elements.wblist (03/10/17)
 --> contains the javascript autocomplete data

html_doc.wblibrarian (03/11/17)
 --> the HTML Librarian document

css_doc.wblibrarian (03/11/17)
 --> The Cascading Style Sheets Librarian document

js_doc.wblibrarian (03/11/17)
 --> The Javascript Librarian document

+-------------------------------------------------------+
+--------------------------Other------------------------+
+-------------------------------------------------------+

You may also contribute in the following areas as well...
    + Documentation
    + Design
    + Support
    + Translation

----

If you are looking for a binary instead, go to http://www.facebook.com/CISoftwareWebAuthor
and download WebAuthor for your operating system.

Currently Tested:
    KDE Neon 5.9.3 Developer Edition
    Kubuntu 16.04.1 LTS

    Windows 10 Pro / Home
    Windows 7 Pro



changes.txt
[03/14/17]
> Improved Code completion
> Added more Javascript code completion terms

[03/13/17]
> Performance tweaks added to WBEditorView's line numbering ( it is still not functional though )
> Line numbers only work on unwrapped text at the moment
> added auto-indentation to WBCodeView
> Code Completion is now functional ( although it is still a bit buggy )

[03/12/17]
> Updated the website links
> Search feature added in the Librarian widget
> Updated Librarian documents
> Added search feature for the "WebAuthor" widget
> Improved search feature
> WBCodeViews should now be used with the WBEditorView widget
> started experimenting with Line numbers in WBEditorWidget

[03/11/17]
> Created the Librarian widget
> Created the first version of the HTML Librarian document
> Created the first version of the CSS Librarian document
> Created the first version of the JS Librarian document

[03/10/17]
> Created changes.txt and readme.txt
> Split the code into multiple files
> Created the WBCodeView class
> Changed the widget type of the editors to WBCodeView
> Added *_elements.wblist files
> Created the WBCodeCompletionModel class [not functional yet]
> Added sliders to the color picker tool
> Color Picker Tool automatically copies the color to the clipboard

:ISSUES:
> Code completion isn't functional yet
> Color picker tool can cause crash


webauthor.h
#ifndef WEBAUTHOR_H
#define WEBAUTHOR_H

#include "wbhtmlsyntax.h"
#include "wbcsssyntax.h"
#include "wbjssyntax.h"
#include "wbcolorlistmodel.h"
#include "wbcodecompletionmodel.h"
#include "wblibrarian.h"
#include "wbeditorview.h"

class WebAuthor : public QMainWindow
{
    Q_OBJECT

public:
    WebAuthor(QWidget *parent = 0);
    ~WebAuthor();

    QClipboard *clipboard;

    //(03/12/17) Text search
    QLineEdit *textSearch;

    //(03/12/17) Editor View
    WBEditorView *markupEditor;
    WBEditorView *cssEditor;
    WBEditorView *jsEditor;

    //(03/11/17) Librarian widget
    WBLibrarian *librarianWidget;

    //WebAuthor widgets
    QMenuBar *menuBar;
    QStatusBar *WBStatusbar;            

    //text areas
    WBCodeView *markupView;
    WBCodeView *cssView;
    WBCodeView *jsView;
    QTextBrowser *previewView;

    //docks
    QDockWidget *previewDock;
    QDockWidget *cssDock;
    QDockWidget *jsDock;

    //extras
    QWidget *colorPickerWidget;
    QColorDialog *colorPickerDialog;
    QListView *colorPickerList;
    WBColorListModel *colorPickerModel;
    QLineEdit *colorPickerOutput;
    QSlider *colorPickerRedValue;
    QSlider *colorPickerGreenValue;
    QSlider *colorPickerBlueValue;

    //document variables
    bool _settingWordWrap = false;
    bool _settingFullscreen = false;

    bool _settingPreviewDockShown = true;
    bool _settingCSSDockShown = false;
    bool _settingJSDockShown = false;
    bool _propertyHadPreviewed = false;
    Qt::DockWidgetArea _settingPreviewDockLocation;
    Qt::DockWidgetArea _settingCSSDockLocation;
    Qt::DockWidgetArea _settingJSDockLocation;

    //(03/12/14) Line number
    bool _settingLineNumberShown = true;

    QString _settingWebAuthorTempPath;

    //file variables
    QString markupCode;
    QString stylesheetCode;
    QString scriptCode;

    //dynamically labeled actions
    QAction *_viewMenuFullscreen;
    QAction *_viewMenuWordwrap;
    QAction *_viewMenuLineNumber; //(03/12/14)
    QAction *_viewMenuPreviewDock;
    QAction *_viewMenuCSSDock;
    QAction *_viewMenuJSDock;

    //function definition(s)
    void setupMenuBar( );
    void applyPreferences( );
    void setPreferences();

    void colorApplySlider();

private:
    QTextBrowser *help;
    QSettings *_userPreferences;
    int currentlyActive = -1;

    bool hasSaved = false;
    QString savedFileName = "";

private slots:
    void newProject();
    void openProject();
    void handleOpenProject(const QString &filename);
    void saveProject();
    void handleSaveProject(const QString &filename);
    void previewPage();
    void exportProject();
    void handleExport(const QString &filename);

    //(03/12/17)
    void showTextSearch();
    void textSearchStringChanged(const QString &searchString);

    void toggleFullscreen();
    void toggleWordwrap();
    void togglePreviewDock();
    void toggleCSSDock();
    void toggleJSDock();

    void showColorPickerWindow( );
    void colorPickerColorSelected( const QColor &color );
    void colorPickerSetText( int red, int green, int blue );
    void colorPickerPickedFromList( const QModelIndex &index );
    void showPicker();
    void colorPickerAdjustRed(int);
    void colorPickerAdjustGreen(int);
    void colorPickerAdjustBlue(int);

    void handlePreviewDockChanges( const Qt::DockWidgetArea &area );
    void handleCSSDockChanges( const Qt::DockWidgetArea &area );
    void handleJSDockChanges( const Qt::DockWidgetArea &area );

    void showGuide();
    void showAbout();
    void showAboutQt();

    void previewDockDisabled( bool s );
    void cssDockDisabled( bool s );
    void jsDockDisabled( bool s );

    //productivity shortcuts
    void switchToMain();
    void switchToStyle();
    void switchToScript();

    //code completion (03/10/17)
    void showCodeCompletion();
    void focusToMarkup();
    void focusToStylesheet();
    void focusToScript();

    //librarian (03/11/17)
    void showLibrarian();

    //line number (03/12/17)
    void toggleLineNumber( );

    void updatePreview( );
};


#endif // WEBAUTHOR_H
